### ISSUE TITLE
Cannot delete IAMRoles with Policies attached.

### ISSUE THREAD

**Issue by preeny95** [rebuy-de/aws-nuke#294]
Hello,

I am running the nuke on a weekly basis in our development account however I cannot remove IAMRoles or Policies. 
```
global - IAMRole - terraform-20181113161447942200000001 - DeleteConflict: Cannot delete entity, must delete policies first.
	status code: 409, request id: c0c0e89f-ebc6-11e8-8118-bf1c41531853
```
And on the policies it states:
```
global - IAMPolicy - arn:aws:iam::XXXXXX:policy/master_14112018_dynamodb_readonly - DeleteConflict: Cannot delete a policy attached to entities.
```
The policies are not attached to anything else and if I manually remove the IAMRole they then successfully delete with the nuke

**Comment by svenwltr** [comment]
Hello @preeny95,

I struggle to reproduce this. Can you give short instructions to create such a setup?

**Comment by preeny95** [comment]
Hi @svenwltr 
```resource "aws_iam_role" "storage_lambda_iam_role" {
  name               = "${terraform.workspace}-storage_lambda_exection"
  assume_role_policy = "${file("${path.module}/templates/lambda_assume_role.json.tpl")}"
}

resource "aws_iam_role_policy_attachment" "role-policy-attachment" {
  role       = "${aws_iam_role.storage_lambda_iam_role.name}"
  count      = "${length(var.storage_lambda_iam_arn)}"
  policy_arn = "${var.storage_lambda_iam_arn[count.index]}"
}
variable "storage_lambda_iam_arn" {
  type        = "list"
  description = "Required polices for the lambda iam role"
  default = ["arn:aws:iam::aws:policy/service-role/AWSLambdaSQSQueueExecutionRole", "arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess", "arn:aws:iam::aws:policy/AWSIoTDataAccess"]
}
```
Thats an example role I am creating in terraform. Which the nuke is unable to delete.
```regions:
- "global"
- "eu-west-1"

resource-types:
  excludes:
  - IAMUser
  - IAMUserAccessKey
  - OpsWorksUserProfile
  - KMSAlias
  - AppStreamImage
  - CloudTrailTrail

# check for further IAM resources we need to preserve 

accounts:
  "XXXXXXXX":
    filters:
        ConfigServiceConfigurationRecorder:
        - "InfraConfigWatcher"
        CloudFormationStack:
        - "StackSet-config-rules-stackset-8be079f8-a2b7-43cd-9eba-59d2fa09549a"
        ConfigServiceDeliveryChannel:
        - "InfraSecurityWatcher"
        ConfigServiceConfigRule:
        - "CloudTrailEnabled"
        - "RDSBackupEnabled"
        - "RDSStorageEncrypted"
        - "RedshiftConfiguredWell"
        - "S3PublicReadCheck"
        - "S3PublicWriteCheck"
        - "SecurityGroupsMustRestrictSSHTraffic"
        IAMRole:
        - "admin"
        - "ApiGatewayPushLogs"
        - "AWSCloudFormationStackSetExecutionRole"
        - "AWSServiceRoleForAmazonGuardDuty"
        - "AWSServiceRoleForECS"
        - "AWSServiceRoleForOrganizations"
        - "AWSServiceRoleForSupport"
        - "AWSServiceRoleForTrustedAdvisor"
        - "guardrails-TrendSecurity"
        - "OrganizationAccountAccessRole"
        - "readonly"
        - "sbg-configservice-role"
        - "SecurityAuditRole"
        - "superadmin"```

Thats also my current nuke config if thats any better to go on.

**Comment by svenwltr** [comment]
@preeny95 I created the resources with Terraform, but I was able to remove the `IAMRole`:

```
global - IAMRole - default-storage_lambda_exection - DeleteConflict: Cannot delete entity, must detach all policies first.
	status code: 409, request id: 8d636809-ee70-11e8-b888-edaaf46a2c89
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AmazonDynamoDBFullAccess - [PolicyName: "AmazonDynamoDBFullAccess", RoleName: "default-storage_lambda_exection"] - triggered remove
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSLambdaSQSQueueExecutionRole - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSLambdaSQSQueueExecutionRole"] - triggered remove
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSIoTDataAccess - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSIoTDataAccess"] - triggered remove

Removal requested: 11 waiting, 1 failed, 125 skipped, 0 finished

global - IAMRole - default-storage_lambda_exection - removed
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AmazonDynamoDBFullAccess - [PolicyName: "AmazonDynamoDBFullAccess", RoleName: "default-storage_lambda_exection"] - waiting
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSLambdaSQSQueueExecutionRole - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSLambdaSQSQueueExecutionRole"] - waiting
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSIoTDataAccess - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSIoTDataAccess"] - waiting

Removal requested: 11 waiting, 0 failed, 125 skipped, 1 finished

global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AmazonDynamoDBFullAccess - [RoleName: "default-storage_lambda_exection", PolicyName: "AmazonDynamoDBFullAccess"] - removed
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSLambdaSQSQueueExecutionRole - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSLambdaSQSQueueExecutionRole"] - removed
global - IAMRolePolicyAttachment - default-storage_lambda_exection -> AWSIoTDataAccess - [RoleName: "default-storage_lambda_exection", PolicyName: "AWSIoTDataAccess"] - removed

Removal requested: 0 waiting, 0 failed, 125 skipped, 12 finished

Nuke complete: 0 failed, 125 skipped, 12 finished.
```

(I removed unrelated resources.)

As you see it displays the same error in the first iteration, but it eventually deletes it and also the policy attachments. Do you see the same behavior, or is the resource actually still there after the *aws-nuke* run?

**Comment by preeny95** [comment]
Hmm I see the same behaviour in the console but all of the resources are still there unfortunately.

**Comment by svenwltr** [comment]
@preeny95 You could turn on the debug output with `-v` and send me the output. *aws-nuke* should strip all sensitive output, but please double check it. To reduce the logging output you can restrict the nuking to the relevant resources with `--target`.

Example:

```
aws-nuke ... -no-dry-run -v -t IAMRole -t IAMRolePolicyAttachment
```

**Comment by preeny95** [comment]
@svenwltr 
```<RoleName>GNS-410-rich-et-a-storage_lambda_exection</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-a-storage_lambda_exection</Arn>\n    <         <CreateDate>2018-11-23T13:24:26Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22AWS%22%3A%22arn%3Aaws%3Aiam%3A%3A219315098781%3Aroot%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%2C%22Condition%22%3A%7B%7D%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJTUN5CA25L3ZI5AIM</RoleId>\n    <         <RoleName>GNS-410-rich-et-a_lambda_dynamodb_readonly_iot_sub_pub</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-a_lambda_dynamodb_readonly_iot_sub_pub</Arn>\n    <         <CreateDate>2018-11-23T13:24:26Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAI3PPKWNDP6IWV33ZW</RoleId>\n    <         <RoleName>GNS-410-rich-et-a_lambda_logs_dynamodb_WNS_readonly</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-a_lambda_logs_dynamodb_WNS_readonly</Arn>\n    <         <CreateDate>2018-11-23T13:24:27Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAIISH2X6TX5II43SKE</RoleId>\n    <         <RoleName>GNS-410-rich-et-a_lambda_logs_SSM_sts_assumerole</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-a_lambda_logs_SSM_sts_assumerole</Arn>\n    <         <CreateDate>2018-11-23T13:24:26Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAIEVFAGGWFXITDHJOO</RoleId>\n    <         <RoleName>GNS-410-rich-et-iam_allow_lambda_invoke</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-iam_allow_lambda_invoke</Arn>\n    <         <CreateDate>2018-11-23T08:30:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAIJFLM65YADDVEIX2Q</RoleId>\n    <         <RoleName>GNS-410-rich-et-state_change_lambda_execution</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-state_change_lambda_execution</Arn>\n    <         <CreateDate>2018-11-23T09:24:32Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJJZY3AOBQN5OXX2OM</RoleId>\n    <         <RoleName>GNS-410-rich-et-state_change_sqs_execution</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-state_change_sqs_execution</Arn>\n    <         <CreateDate>2018-11-23T09:24:32Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJF2MI5LH33XNILZHC</RoleId>\n    <         <RoleName>GNS-410-rich-et-storage_lambda_exection</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et-storage_lambda_exection</Arn>\n    <         <CreateDate>2018-11-23T08:30:13Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22AWS%22%3A%22arn%3Aaws%3Aiam%3A%3A219315098781%3Aroot%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%2C%22Condition%22%3A%7B%7D%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAI65QCL7MI2AOCWWZY</RoleId>\n    <         <RoleName>GNS-410-rich-et_lambda_dynamodb_readonly_iot_sub_pub</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et_lambda_dynamodb_readonly_iot_sub_pub</Arn>\n    <         <CreateDate>2018-11-23T08:30:13Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAITEZ3JMGTHREKHQ5S</RoleId>\n    <         <RoleName>GNS-410-rich-et_lambda_logs_dynamodb_WNS_readonly</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et_lambda_logs_dynamodb_WNS_readonly</Arn>\n    <         <CreateDate>2018-11-23T08:30:13Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAIJ45MRRKNHGT2E4A2</RoleId>\n    <         <RoleName>GNS-410-rich-et_lambda_logs_SSM_sts_assumerole</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-rich-et_lambda_logs_SSM_sts_assumerole</Arn>\n    <         <CreateDate>2018-11-23T08:30:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJUHI5B6DCLZMM432W</RoleId>\n    <         <RoleName>GNS-410-storage_lambda_exection</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410-storage_lambda_exection</Arn>\n    <         <CreateDate>2018-11-19T15:54:40Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJNG5QGZ3KA6UH6DHW</RoleId>\n    <         <RoleName>GNS-410a-iam_allow_lambda_invoke</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a-iam_allow_lambda_invoke</Arn>\n    <         <CreateDate>2018-11-20T11:35:15Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJGLKKRBOASBYQNCCK</RoleId>\n    <         <RoleName>GNS-410a-state_change_lambda_execution</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a-state_change_lambda_execution</Arn>\n    <         <CreateDate>2018-11-21T11:07:54Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAIWNIDPFWWKO5Y3QGG</RoleId>\n    <         <RoleName>GNS-410a-state_change_sqs_ex\r\n    < 10f5\r\n    < ecution</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a-state_change_sqs_execution</Arn>\n    <         <CreateDate>2018-11-21T11:04:21Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJO7NG6ULBJFD347QC</RoleId>\n    <         <RoleName>GNS-410a-storage_lambda_exection</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a-storage_lambda_exection</Arn>\n    <         <CreateDate>2018-11-20T11:35:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22AWS%22%3A%22arn%3Aaws%3Aiam%3A%3A219315098781%3Aroot%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%2C%22Condition%22%3A%7B%7D%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJ3OIJFYI33JTACHQ6</RoleId>\n    <         <RoleName>GNS-410a_lambda_dynamodb_readonly_iot_sub_pub</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a_lambda_dynamodb_readonly_iot_sub_pub</Arn>\n    <         <CreateDate>2018-11-20T11:35:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAI6K3GOSNC6AOGDHWM</RoleId>\n    <         <RoleName>GNS-410a_lambda_logs_dynamodb_WNS_readonly</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a_lambda_logs_dynamodb_WNS_readonly</Arn>\n    <         <CreateDate>2018-11-20T11:35:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Sid%22%3A%22%22%2C%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22Service%22%3A%22lambda.amazonaws.com%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJD44DDA4T6EWLWF2Q</RoleId>\n    <         <RoleName>GNS-410a_lambda_logs_SSM_sts_assumerole</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410a_lambda_logs_SSM_sts_assumerole</Arn>\n    <         <CreateDate>2018-11-20T11:35:14Z</CreateDate>\n    <       </member>\n    <       <member>\n    <         <Path>/</Path>\n    <         <AssumeRolePolicyDocument>%7B%22Version%22%3A%222012-10-17%22%2C%22Statement%22%3A%5B%7B%22Effect%22%3A%22Allow%22%2C%22Principal%22%3A%7B%22AWS%22%3A%22arn%3Aaws%3Aiam%3A%3A219315098781%3Aroot%22%7D%2C%22Action%22%3A%22sts%3AAssumeRole%22%2C%22Condition%22%3A%7B%7D%7D%5D%7D</AssumeRolePolicyDocument>\n    <         <MaxSessionDuration>3600</MaxSessionDuration>\n    <         <RoleId>AROAJVCNO7KRNA3ORRUM2</RoleId>\n    <         <RoleName>GNS-410_lambda_dynamodb_readonly_iot_sub_pub</RoleName>\n    <         <Arn>arn:aws:iam::219315098781:role/GNS-410_lambda_dynamodb_readonly_iot_sub_pub</Arn>\n    <         <CreateDate>2018-11-19T15:54:40Z</CreateDate>\n 
``` 
Theres quite abit snippet if thats any help to you. If needs be I can get the full jenkins output to you.

**Comment by preeny95** [comment]
We've also just had to manually remove the policies again today for the nuke to work removing the IAMRoles.

**Comment by svenwltr** [comment]
Hello @preeny95. Sorry for the late response. Your logs looks quite stripped. Can you maybe put it into a [gist](https://gist.github.com/)? If you worry about sensitive data, you can send it directly to me via s.walter@rebuy.com

**Comment by preeny95** [comment]
Hi @svenwltr I've sent you the link to the gist via email

**Comment by svenwltr** [comment]
Hello @preeny95,

I noticed from your logs that, *aws-nuke* tries to remove IAM Roles, but it does not try to remove the Policies. First, I thought this might be a pagination issue, since there are a lot of roles, but then I notices this:

```
Listing IAMRolePolicyAttachment failed. Please report this to https://github.com/rebuy-de/aws-nuke/issues/new.
    !!! AccessDenied: User: arn:aws:sts::xxx:assumed-role/admin/vault-approle-admin-xxx-yyy is not authorized to perform: iam:ListAttachedRolePolicies on resource: role admin
    !!!     status code: 403, request id: 1234
```

So it seems to be an permission problem. Since you filtered the `admin` role, I guess you actually do not want to delete it, but it looks like this error is not handled properly and *aws-nuke* fails to list any attachments.

I will fix this soon.

**Comment by preeny95** [comment]
Hi @svenwltr 

Thank you for looking into this!

**Comment by preeny95** [comment]
Hi @svenwltr 
We have ran the nuke this morning and are getting the same error message and IAM roles still not being removed.
```
time="2018-12-10T05:46:37Z" level=error msg="Listing IAMLoginProfile failed. Please report this to https://github.com/rebuy-de/aws-nuke/issues/new.\n    !!! AccessDenied: User: arn:aws:sts::XXXXXXX:assumed-role/admin/vault-approle-admin-XXXXXX-3159 is not authorized to perform: iam:GetLoginProfile on resource: user vault\n    !!! \tstatus code: 403, request id: f564dd5d-fc3e-11e8-9d97-dfa6af9addf5"
```

**Comment by svenwltr** [comment]
@preeny95 I see. This seems to be the same problem with another resource. I will take care.

**Comment by svenwltr** [comment]
Should be fixed in #309
